### PR TITLE
Add conversion from Tensor<double> to ndarray with Pypp

### PR DESCRIPTION
**Pull request by osheamonn**
## Proposed changes

Add conversion for arbitrary rank Tensors to NumPy ndarrays. 

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [ ] Code
- [x] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- Code has documentation and unit tests
- Private member variables have a trailing underscore
- Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- File lists in CMake are alphabetical
- Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- Mark objects `const` whenever possible
- Almost always `auto`, except with expression templates, i.e. `DataVector`
- All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- Prefix commits addressing PR requests with `fixup`. These commits are flagged
  by Travis so we do not accidentally merge them.
- Include what you use, prefer forward declarations in hpp files.
- Explicitly make numbers floating point, e.g. `2.` or `2.0` over `2`
- Use `Tensor`'s non-member get if the indices are constant expressions, e.g.
  `get<1>(tensor)`

### CODE REVIEW

File: cmake/FindNumPy.cmake
Comment (mar-celine) [inline cmake/FindNumPy.cmake]: remove extra white space before __values

File: cmake/FindNumPy.cmake
Comment (mar-celine) [inline cmake/FindNumPy.cmake]: SCREAMING_SNAKE_CASE (for __result, etc)

File: cmake/FindNumPy.cmake
Comment (mar-celine) [inline cmake/FindNumPy.cmake]: "Failed to retrieve" instead of "Requested"

File: tests/Unit/Pypp/PyppFundamentals.hpp
Comment (gsb76) [inline tests/Unit/Pypp/PyppFundamentals.hpp]: Is there a nondeprecated api?

File: tests/Unit/Pypp/PyppFundamentals.hpp
Comment (mar-celine) [inline tests/Unit/Pypp/PyppFundamentals.hpp]: why not use std::array?

File: tests/Unit/Pypp/PyppFundamentals.hpp
Comment (mar-celine) [inline tests/Unit/Pypp/PyppFundamentals.hpp]: rename to npy_array or similar

File: tests/Unit/Pypp/PyppFundamentals.hpp
Comment (mar-celine) [inline tests/Unit/Pypp/PyppFundamentals.hpp]: same name as what was suggested to be named npy_array above to be consistent

File: tests/Unit/Pypp/PyppFundamentals.hpp
Comment (mar-celine) [inline tests/Unit/Pypp/PyppFundamentals.hpp]: const auto value = static_cast<const double*>

File: tests/Unit/Pypp/PyppFundamentals.hpp
Comment (mar-celine) [inline tests/Unit/Pypp/PyppFundamentals.hpp]: why cant you use the one inside tensor structure?

File: tests/Unit/Pypp/PyppFundamentals.hpp
Comment (mar-celine) [inline tests/Unit/Pypp/PyppFundamentals.hpp]: pointer can be const, do this in other commit

File: tests/Unit/Pypp/PyppFundamentals.hpp
Comment (mar-celine) [inline tests/Unit/Pypp/PyppFundamentals.hpp]: do not default construct

File: tests/Unit/Pypp/Test_Pypp.cpp
Comment (mar-celine) [inline tests/Unit/Pypp/Test_Pypp.cpp:131]: add tnsr::aia to also test this (or something like this)

**Review by mar-celine — CHANGES_REQUESTED** [review]

File: tests/Unit/Pypp/PyppFundamentals.hpp
Comment (osheamonn) [reply tests/Unit/Pypp/PyppFundamentals.hpp]: Because constructing the PyArray requires a c-style array of the number of components in each dimension. I could make this a std::array but I'd only be doing so to call `.data()` on it anyway so I thought it'd be clearer to keep it as a c-style array since that what the function below takes. 

Is there a way to inline the construction of the c-style array in the call to PyArray_SimpleNew? Since it's only used there it might make it clearer.

File: tests/Unit/Pypp/PyppFundamentals.hpp
Comment (osheamonn) [reply tests/Unit/Pypp/PyppFundamentals.hpp]: I didn't see it in Structure? There's a tensor_index from collapsed function, and a collapsed_to_storage which takes into account the symmetry of the tensor which we don't want to do here.

File: tests/Unit/Pypp/PyppFundamentals.hpp
Comment (kidder) [reply tests/Unit/Pypp/PyppFundamentals.hpp]: @nilsdeppe do you have an opinion?

File: tests/Unit/Pypp/PyppFundamentals.hpp
Comment (kidder) [reply tests/Unit/Pypp/PyppFundamentals.hpp]: can we use IndexIterator?

File: tests/Unit/Pypp/PyppFundamentals.hpp
Comment (nilsdeppe) [reply tests/Unit/Pypp/PyppFundamentals.hpp]: I am in favor of using `std::array`. Inside the H5 interface C-style arrays are "needed" and there I used `std::array` and called `.data()`. In general, https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Res-stack

File: tests/Unit/Pypp/PyppFundamentals.hpp
Comment (osheamonn) [reply tests/Unit/Pypp/PyppFundamentals.hpp]: I think I can use IndexIterator but I have to change two things about index: 
- I've to be able to get a std::array<size_t> from an Index
- I've to allow the Dim parameter of index to take values up to 4 (or more if we have larger rank tensors).

File: tests/Unit/Pypp/PyppFundamentals.hpp
Comment (mar-celine) [inline tests/Unit/Pypp/PyppFundamentals.hpp]: remove clang tidy comment

File: tests/Unit/Pypp/PyppFundamentals.hpp
Comment (mar-celine) [inline tests/Unit/Pypp/PyppFundamentals.hpp]: factor this out and rename it to something helpful

File: tests/Unit/Pypp/PyppFundamentals.hpp
Comment (mar-celine) [inline tests/Unit/Pypp/PyppFundamentals.hpp]: this name is backwards

File: tests/Unit/Pypp/PyppPyTests.py
Comment (nilsdeppe) [inline tests/Unit/Pypp/PyppPyTests.py]: Experience from SpEC has shown that it's good to avoid the words "these" and "this" in comments in favor of people specific. I'd prefer "The functions scalar, vector, tnsr* are used..."

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
Could you squash and rebase this on top of #553 after that one has seen the changes I requested there and was rebased on develop?

File: src/DataStructures/Index.hpp
Comment (nilsdeppe) [inline src/DataStructures/Index.hpp]: move back to cpp file, add `4` to list, leave tests :+1:

File: tests/Unit/Pypp/SetupLocalPythonEnvironment.hpp
Comment (nilsdeppe) [inline tests/Unit/Pypp/SetupLocalPythonEnvironment.hpp]: add comment about why python sucks....

File: tests/Unit/Pypp/PyppFundamentals.hpp
Comment (nilsdeppe) [inline tests/Unit/Pypp/PyppFundamentals.hpp]: `auto*`

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
squash, rebase, run through ASAN

**Comment by osheamonn** [comment]
ASAN gives no complaints.

**Review by nilsdeppe — APPROVED** [review]

**Comment by osheamonn** [comment]
Hold off on merging, may have spoken too soon with regards to ASAN...

**Review by kidder — DISMISSED** [review]

**Comment by kidder** [comment]
@gsb76 you need to approve this (or request new changes)

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
blocking as per @osheamonn's request

**Comment by osheamonn** [comment]
False alarm, passes ASAN. A (future) dependent PR wasn't passing it so I was worried that this PR wasn't either, but I've confirmed that it's only the future changes which break it.

**Review by nilsdeppe — DISMISSED** [review]
Approving after Eamonn's mini scare :) Thanks for being thorough! :+1:

**Review by mar-celine — DISMISSED** [review]

**Comment by osheamonn** [comment]
At least now I know what happens if a function that shouldn't be noexcept is marked as such.

**Comment by nilsdeppe** [comment]
Did std::terminate get invoked and then the code called Parallel::abort? This is what we would like to have happen :)

**Comment by osheamonn** [comment]
Yeah but with Python code you seem to get leaks if that happens, so no noexcept here!

**Comment by nilsdeppe** [comment]
Ah RIP, yea I guess it wouldn't have freed the memory before terminating in that case... Well, at least now we know :)

**Comment by nilsdeppe** [comment]
It looks like Mike uses conda on the macOS travis builds: https://github.com/moble/quaternion/blob/master/.travis.yml
Let's give that a try to see if it fixes these issues...

**Comment by nilsdeppe** [comment]
What I mean is, we should try grabbing miniconda, install numpy, and using that on the macOS builds. That should not make build times longer and provide stability

File: .travis.yml
Comment (nilsdeppe) [inline .travis.yml]: It looks like Travis didn't actually do anything with this update, as in I can't find an associated build. It does look like it needs a whole bunch of semicolons, though, which might be the issue.

**Comment by nilsdeppe** [comment]
Can you try adding `export CMAKE_PREFIX_PATH=$HOME/miniconda:$CMAKE_PREFIX_PATH;`? It looks like CMake is still picking up the system python installation :disappointed:

**Review by nilsdeppe — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Pull request merged**
